File: App Factory/IAIconDrop.m
#import "IAIconDrop.h"


@implementation IAIconDrop {}

-(id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder: aDecoder];
    [self registerForDraggedTypes: @[NSURLPboardType]];
    return self;
}

- (BOOL)performDragOperation:(id <NSDraggingInfo>)sender
{
    NSPasteboard *pboard = [sender draggingPasteboard];

    if ( [[pboard types] containsObject:NSURLPboardType] ) {
        self.iconPath = [NSURL URLFromPasteboard:pboard];

        [self.iconLabel setHidden:YES];
        // Show the selected icon image
        NSImage *newImage = [[NSImage alloc] initWithContentsOfURL:self.iconPath];
        [self setImage:newImage];
        [self setNeedsDisplay:YES];
    }

    return YES;
}

- (NSDragOperation)draggingEntered:(id <NSDraggingInfo>)sender
{
    NSPasteboard *pboard;
    NSDragOperation sourceDragMask;

    sourceDragMask = [sender draggingSourceOperationMask];
    pboard = [sender draggingPasteboard];

    if ( [[pboard types] containsObject:NSURLPboardType] ) {
        if (sourceDragMask & NSDragOperationLink) {
            return NSDragOperationLink;
        } else if (sourceDragMask & NSDragOperationCopy) {
            return NSDragOperationCopy;
        }
    }
    return NSDragOperationNone;
}

@end

File: App Factory/IAIconDrop.swift
import Foundation
import Cocoa

class IAIconDrop: NSImageView {
    @IBOutlet var iconLabel: NSTextField!
    var iconPath: NSURL!
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        registerForDraggedTypes([NSURLPboardType])
    }
    
    override func performDragOperation(sender: NSDraggingInfo) -> Bool {
        let pboard = sender.draggingPasteboard()
        if !pboard.types!.contains(NSURLPboardType) {
            return true
        }
        
        self.iconLabel.hidden = true
        self.iconPath = NSURL(fromPasteboard: pboard)
        self.image = NSImage(contentsOfURL: iconPath)
        self.needsDisplay = true
        return true
    }
    
    override func draggingEntered(sender: NSDraggingInfo) -> NSDragOperation {
        let sourceDragMask = sender.draggingSourceOperationMask()
        let pboard = sender.draggingPasteboard()
        
        if pboard.types!.contains(NSURLPboardType) {
            if sourceDragMask.contains(NSDragOperation.Link) {
                return NSDragOperation.Link
            } else if sourceDragMask.contains(NSDragOperation.Copy) {
                return NSDragOperation.Copy
            }
        }
        
        return NSDragOperation.None
    }
}
